r2dbc-spi/src/main/java/io/r2dbc/spi/Result.java

```diff
@@ -49,4 +54,124 @@
      */
     <T> Publisher<T> map(BiFunction<Row, RowMetadata, ? extends T> mappingFunction);
 
+    /**
+     * Returns a filtered variant of the {@link Result} to selectively consume result segments matching {@link Predicate filter predicate}.
+     * <p>The returned {@link Result} is a potentially reduced view of the underlying {@link Result} to filter out unwanted result segments.  For example, filtering {@link Message} segments of
+     * {@link Message.Severity#ERROR error severity} from the result
+     * lets the result complete without an error.
+     *
+     * @param filter the non-interfering and stateless {@link Predicate} to apply to each element to determine if it should be included
+     * @return a filtered {@link Result}
+     * @throws IllegalArgumentException if {@code filter} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    Result filter(Predicate<Segment> filter);
+
+    /**
+     * Returns a mapping of the {@link Segment result segments} that are the results of a query against a database.  May be empty if the query did not return any segments.  A {@link Segment} can be
+     * only considered valid within a {@link Function mapping function} callback.
+     * <p>Consuming result {@link Segment segments} does not emit {@link Subscriber#onError(Throwable) error signals} from {@link Message#severity() error segments}.  Translation of error segments
+     * is subject to the {@code mappingFunction}.
+     *
+     * @param mappingFunction the {@link Function} that maps a {@link Segment} a to a value
+     * @param <T>             the type of the mapped value
+     * @return a mapping of the segments that are the results of a query against a database
+     * @throws IllegalArgumentException if {@code mappingFunction} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    <T> Publisher<T> flatMap(Function<Segment, ? extends Publisher<T>> mappingFunction);
```

Comment:
I wonder if we can extend further on the monadic method name convention on `M<T>` with (simplified):

- `M<T>.map(F<T, U>): M<U>`
- `M<T>.flatMap(F<T, M<U>>): M<U>`

Currently, the types aren't exactly what one might expect from reading the method names `map/flatMap`. With this new `Segment` type, I wonder if a `Result` is really some sort of `M<Segment>` and the `map()` operation should receive a `F<Segment, T>` function instead of the specific one that is strictly tied to the `Data` subtype of `Segment`?

Reply:
Strictly speaking, `Result` becomes `Publisher<Segment>` via `map(…)`. It semantically translates to `filter(Data.class::instance).cast(Data.class).map(data -> mappingFunction(data.row(), data.metadata())`.

The intent behind `Row` (and everything associated with resources) is that we do not want to these to leave the `Result` closure via `Result.map(…)` (nor `flatMap(…)`) as their state shall be managed by `Result`. 

Another way out could be renaming `map` to `mapRow(…)` but that would have a bigger impact.